test/integration/admin/api/services_controller_test.rb

```diff
@@ -3,69 +3,126 @@
+        get admin_api_service_path(service, format: format)
+        assert_response :ok
+        assert response.body.include?('deployment_option')
+        assert response.body.include?(service.deployment_option)
+      end
     end
-  end
 
-  def test_show
-    service = master_account.default_service
-    %i[xml json].each do |format|
-      get admin_api_service_path(service, format: format)
+    test 'index works for SaaS but it is unauthorized for Master On-prem' do
+      ThreeScale.stubs(master_on_premises?: false)
+      get admin_api_services_path
       assert_response :ok
-      assert response.body.include?('deployment_option')
-      assert response.body.include?(service.deployment_option)
+
+      ThreeScale.stubs(master_on_premises?: true)
+      get admin_api_services_path
+      assert_response :forbidden
     end
   end
 
-  test 'GET index works for Saas' do
-    get admin_api_services_path
-    assert_response :ok
-  end
+  class TenantHostTest < ActionDispatch::IntegrationTest
+    setup do
+      @provider = FactoryBot.create(:provider_account)
+      host! provider.admin_domain
+      @service = FactoryBot.create(:service, account: provider)
+      Account.any_instance.stubs(can_create_service?: true)
+    end
 
-  test 'GET index is unauthorized for Master On-prem' do
-    ThreeScale.stubs(master_on_premises?: true)
-    get admin_api_services_path
-    assert_response :forbidden
-  end
+    attr_reader :provider, :service
 
-  class WithProviderKey < ActionDispatch::IntegrationTest
+    test 'delete with api_key' do
+      assert_change(of: -> { service.reload.deleted? }, from: false, to: true) do
+        delete admin_api_service_path(service, provider_key: provider.api_key)
+        assert_response :ok
+      end
+    end
 
-    def test_delete
-      provider = FactoryBot.create(:provider_account)
-      host! provider.admin_domain
-      service = FactoryBot.create(:service, account: provider)
-      refute service.deleted?
-      delete admin_api_service_path(service, provider_key: provider.api_key)
-      assert_response :ok
-      assert service.reload.deleted?
+    test 'create' do
+      assert_difference(Service.method(:count)) do
+        post admin_api_services_path(access_token: access_token_value, format: :json), permitted_params.merge(forbidden_params).merge({system_name: 'the-system-name'})
+        assert_response :created
+      end
+      assert(@service = provider.services.find_by(id: JSON.parse(response.body).dig('service', 'id')))
+      assert_persisted_right_params
+      assert_equal 'the-system-name', service.system_name
+    end
+
+    test 'update' do
+      put admin_api_service_path(service, access_token: access_token_value, format: :json), permitted_params.merge(forbidden_params).merge({system_name: 'the-system-name'})
+      assert_response :success
+      service.reload
+      assert_persisted_right_params
+      assert_not_equal 'the-system-name', service.system_name
```

Comment:
Gui suggested instead:
`assert_equal 'old-system-name', service.reload.system_name`
I said:
> i think both then... because if the old value is equal to the new one... bad test 😄